modules/couchdb/collect.go

```diff
@@ -0,0 +1,243 @@
+package couchdb
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/netdata/go.d.plugin/pkg/stm"
+	"github.com/netdata/go.d.plugin/pkg/web"
+)
+
+const (
+	urlPathActiveTasks   = "/_active_tasks"
+	urlPathOverviewStats = "/_node/%s/_stats"
+	urlPathSystemStats   = "/_node/%s/_system"
+	urlPathDatabases     = "/_dbs_info"
+
+	httpStatusCodePrefix    = "couchdb_httpd_status_codes_"
+	httpStatusCodePrefixLen = len(httpStatusCodePrefix)
+)
+
+func (cdb *CouchDB) collect() (map[string]int64, error) {
+	ms := cdb.scrapeCouchDB()
+	if ms.empty() {
+		return nil, nil
+	}
+
+	collected := make(map[string]int64)
+	cdb.collectNodeStats(collected, ms)
+	cdb.collectSystemStats(collected, ms)
+	cdb.collectActiveTasks(collected, ms)
+	cdb.collectDBStats(collected, ms)
+
+	return collected, nil
+}
+
+func (CouchDB) collectNodeStats(collected map[string]int64, ms *cdbMetrics) {
+	if !ms.hasNodeStats() {
+		return
+	}
+
+	for metric, value := range stm.ToMap(ms.NodeStats) {
+		collected[metric] = value
+		if strings.HasPrefix(metric, httpStatusCodePrefix) {
+			code := metric[httpStatusCodePrefixLen:]
+			collected["couchdb_httpd_status_codes_"+string(code[0])+"xx"] += value
+		}
+	}
+}
+
+func (CouchDB) collectSystemStats(collected map[string]int64, ms *cdbMetrics) {
+	if !ms.hasNodeSystem() {
+		return
+	}
+
+	for metric, value := range stm.ToMap(ms.NodeSystem) {
+		collected[metric] = value
+	}
+
+	collected["peak_msg_queue"] = findMaxMQSize(ms.NodeSystem.MessageQueues)
+}
+
+func (CouchDB) collectActiveTasks(collected map[string]int64, ms *cdbMetrics) {
+	collected["active_tasks_indexer"] = 0
+	collected["active_tasks_database_compaction"] = 0
+	collected["active_tasks_replication"] = 0
+	collected["active_tasks_view_compaction"] = 0
+
+	if !ms.hasActiveTasks() {
+		return
+	}
+
+	for _, task := range ms.ActiveTasks {
+		collected["active_tasks_"+task.Type]++
+	}
+}
+
+func (cdb *CouchDB) collectDBStats(collected map[string]int64, ms *cdbMetrics) {
+	if !ms.hasDBStats() {
+		return
+	}
+
+	for _, dbStats := range ms.DBStats {
+		if dbStats.Error != "" {
+			cdb.Warning("database '", dbStats.Key, "' doesn't exist")
+			continue
+		}
+		merge(collected, stm.ToMap(dbStats.Info), "db_"+dbStats.Key)
+	}
+
+	for metric, value := range stm.ToMap(ms.DBStats) {
+		collected[metric] = value
+	}
+}
+
+func (cdb CouchDB) scrapeCouchDB() *cdbMetrics {
+	ms := &cdbMetrics{}
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+	go func() { defer wg.Done(); cdb.scrapeNodeStats(ms) }()
+
+	wg.Add(1)
+	go func() { defer wg.Done(); cdb.scrapeSystemStats(ms) }()
+
+	wg.Add(1)
+	go func() { defer wg.Done(); cdb.scrapeActiveTasks(ms) }()
+
+	wg.Add(1)
+	go func() { defer wg.Done(); cdb.scrapeDBStats(ms) }()
```

Comment:
do we treat "" databases option as all databases and none? If none then we need to start `cdb.scrapeDBStats` only if len cdb.database > 0.